HW2/P4/driver.py

```diff
@@ -14,16 +14,47 @@
 import filtering
 from timer import Timer
 import threading
+from multiprocessing.pool import ThreadPool
 
 def py_median_3x3(image, iterations=10, num_threads=1):
-    ''' repeatedly filter with a 3x3 median '''
+    ''' repeatedly filter with a 3x3 median and multiple threads '''
     tmpA = image.copy()
     tmpB = np.empty_like(tmpA)
-
-    for i in range(iterations):
-        filtering.median_3x3(tmpA, tmpB, 0, 1)
-        # swap direction of filtering
-        tmpA, tmpB = tmpB, tmpA
+    events = []
+    threads = []
+
+    # events[] has an event for each thread i finishing iteration n. It looks like:
+    # [thr0 iter1, thr0 iter2, ..., thr0 iter k, thr1 iter1, ..., thrN iterK]
+    # This way, every thread can see when a particular thread has finished a
+    # particular iteration.
+    #
+    # There are more details in P4.txt
+
+    def process_image(threadnum, tmpA, tmpB, offset=num_threads):
+        ''' do filtering for the threadnum-th thread '''
+        for iter in range(iterations):
+            # We don't have to check for this thread finishing iteration i-1 because
+            # it does each of its iterations in sequence.
+            if iter > 0:
+                if threadnum != 0:
+                    events[(threadnum-1)*iterations+iter-1].wait()
+                if threadnum != num_threads-1:
+                    events[(threadnum+1)*iterations+iter-1].wait()
```

Comment:
Code still really easy to follow. To avoid the corner case here you could have set num_threads + 2 events per iteration and having the first and the last event always set (they will correspond to the two borders)